doc/lightning-fundpsbt.7.md

```diff
@@ -0,0 +1,67 @@
+lightning-fundpsbt -- Command to populate PSBT inputs from the wallet
+================================================================
+
+SYNOPSIS
+--------
+
+**fundpsbt** *satoshi* *feerate* \[*minconf* *reserve*\]
+
+DESCRIPTION
+-----------
+
+`fundpsbt` is a low-level RPC command which creates a PSBT using unreserved
+inputs in the wallet, optionally reserving them as well.
+
+*satoshi* is the minimum satoshi value of the output(s) needed (or the
+string "all" meaning use all unreserved inputs).  If a value, it can
+be a whole number, a whole number ending in *sat*, a whole number
+ending in *000msat*, or a number with 1 to 8 decimal places ending in
+*btc*.
```

Comment:
We should probably mention that:

* `fundpsbt` factors in the fees needed for each input it proposes to add.
* the given `satoshi` value should include the fee for the outputs, plus the fee for the common parts (`nLockTime`, `nVersion` etc) of the transaction.

The latter point implies that plugins need access to the fee estimates.  While plugins can call the `bcli` `estimatefees` directly, I think it is best if `lightningd/chaintopology` also exposes the actual feerates it would use, especially since `estimatefees` is allowed to return `null` and `chaintopology`, to my understanding, papers over that case. (I think? @darosior )

Alternately, since e.g. Liquid apparently has explicit fee outputs, maybe we can add another parameter, `outputs`, which is an array of strings or `null`s:

* An entry can be an address, in which case `fundpsbt` allocates fees enough for that address.
* An entry can be any of the following strings, indicating `fundpsbt` allocates fees enough for that *kind* of address:
  * `p2pkh`
  * `p2pk` - or not who uses this nowadays
  * `p2sh`
  * `p2wpkh`
  * `p2wsh`
  * `op-return-#` where # is a number from 0 to 64 (or whatever the `OP_RETURN` max is).
* An entry can be `null`, in which case `fundpsbt` will allocate fees for the largest known valid address type `scriptPubKey`..

Then `fundpsbt` can consider whether there is some excess and see if it is large enough to (1) pay for its existence as a change output and (2) remains large enough to keep itself above the dust limit.  If not, it adds the excess to fees (which it adds as an explicit output for Liquid txes).

With this `outputs` parameter we can also have `fundpsbt` consider the common parts of the transaction (`nLockTime` and friends).  This removes any remaining reasons that a robust plugin would want to loop around `fundpsbt`.

Reply:
> The latter point implies that plugins need access to the fee estimates. While plugins can call the bcli estimatefees directly, I think it is best if lightningd/chaintopology also exposes the actual feerates it would use, especially since estimatefees is allowed to return null and chaintopology, to my understanding, papers over that case. (I think? @darosior )

Right, better to use the `feerate` command which returns a human error in case of estimation failure and smoothed estimation otherwise.

`estimatefees` could fail we allow it to, while `feerates` would still return a result based on our cached history.

Reply:
Ah, right, `feerates`.  Will go document it for now.

Reply:
> We should probably mention that:
> 
>     * `fundpsbt` factors in the fees needed for each input it proposes to add.
> 
>     * the given `satoshi` value should include the fee for the outputs, plus the fee for the common parts (`nLockTime`, `nVersion` etc) of the transaction.

The next paragraph does exactly this:

You calculate the value by starting with the amount you want to pay
and adding the fee which will be needed to pay for the base of the
transaction plus that output, and any other outputs and inputs you
will add to the final transaction.

> With this `outputs` parameter we can also have `fundpsbt` consider the common parts of the transaction (`nLockTime` and friends). This removes any remaining reasons that a robust plugin would want to loop around `fundpsbt`.

Indeed.  My draft `txprepare-as-plugin` patch does exactly this, and it's pretty simple (now we have the helpers in place!).

Reply:
> Indeed. My draft `txprepare-as-plugin` patch does exactly this, and it's pretty simple (now we have the helpers in place!).

Is the code in `txprepare` or in `fundpsbt`?  I cannot find `txprepare-as-plugin` branch in your repo or in the ElementsProject repo.  My concern is that I believe you mentioned deprecation of `txprepare`, and in any case `fundchannel`/`multifundchannel` will want to move away from `txprepare` anyway.  So a command which wraps `fundpsbt` and handles as well the common and output weight fees would be quite convenient; if `txprepare` will be deprecated later, it would be useful for both `withdraw` and `fundchannel` (and their `multi` versions) to use that rather than the existing base `fundpsbt`.